{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from sklearn.cluster import KMeans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from ghmm import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "grfKMeans = pickle.load( open( \"grfKMeans.model\", \"rb\" ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "grfDataset = pickle.load( open( \"grfDataset_standing\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1994, 10)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grfDataset.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([40,  6, 40, 40,  6, 40, 40,  6,  6, 40, 60,  6, 40, 36, 36, 40, 40,\n",
       "        6,  6,  6], dtype=int32)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grfKMeans.predict(grfDataset[1800:1800+20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "i = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "N = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "l = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seqList = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seqLengths = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "while True:\n",
    "    block = grfDataset[i:i+N]\n",
    "    seq = grfKMeans.predict(block)\n",
    "    l = seq.size\n",
    "    seqList.append(seq)\n",
    "    \n",
    "    count = count + 1\n",
    "    if l != N:\n",
    "        break\n",
    "    i = i + N\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#seqList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 123"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "npA = np.random.dirichlet((1, 2, 3, 4, 5),size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "npB = np.random.dirichlet((1,2,3,4,5,6,7,8,9,10,\n",
    "                           11,12,13,14,15,16,17,18,19,20,\n",
    "                           21,22,23,24,25,26,27,28,29,30,\n",
    "                           31,32,33,34,35,36,37,38,39,40,\n",
    "                           41,42,43,44,45,46,47,48,49,50,\n",
    "                           51,52,53,54,55,56,57,58,59,60,\n",
    "                           61,62,63,64),size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5, 5)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "npA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5, 64)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "npB.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "A=npA.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "B=npB.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Pi=[1.0/5] * 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sigma = IntegerRange(0,64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hmmModel = HMMFromMatrices(sigma, DiscreteDistribution(sigma), A, B, Pi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trainSet = SequenceSet(sigma, seqList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hmmModel.baumWelch(trainSet, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "testSeq=EmissionSequence(sigma, [40,  6, 40, 40,  6, 40, 40,  6,  6, 40, 60,  6, 40, 36, 36, 40, 40,\n",
    "        6,  6,  6] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 0, 0, 0, 0, 0], -31.77359208832549)\n"
     ]
    }
   ],
   "source": [
    "print(hmmModel.viterbi(testSeq))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# 參考: \n",
    "http://ghmm.sourceforge.net/documentation.html\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#print(hmmModel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#pickle.dump(hmmModel, open(\"ghmmModel_standing\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#help('ghmm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#hmmModel.cmodel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "testSeq1=EmissionSequence(sigma, [ 5, 22, 16,  2, 59, 10, 20, 20, 59,  4, 56, 43, 26, 57,  5, 54, 54,\n",
    "        5, 16, 44] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GHMM ghmm.py:148 - viterbi.c:ghmm_dmodel_viterbi(328): state_seq_index = -1, t = 19\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([-1], 1.0)\n"
     ]
    }
   ],
   "source": [
    "print(hmmModel.viterbi(testSeq1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# backup and restore functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def myGHMMDump(sigma_, hmmModel_, fileName_):\n",
    "    A_=hmmModel_.asMatrices()[0]\n",
    "    B_=hmmModel_.asMatrices()[1]\n",
    "    Pi_=hmmModel_.asMatrices()[2]\n",
    "    \n",
    "    HMM_PARAMS_=[sigma_, A_, B_, Pi_]\n",
    "    pickle.dump(HMM_PARAMS_, open(fileName_, \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def myGHMMLoad(fileName_):\n",
    "    HMM_PARAMS_ = pickle.load(open(fileName_, \"rb\"))\n",
    "    sigma_=HMM_PARAMS_[0]\n",
    "    A_=HMM_PARAMS_[1]\n",
    "    B_=HMM_PARAMS_[2]\n",
    "    Pi_=HMM_PARAMS_[3]\n",
    "    \n",
    "    hmmModel_ = HMMFromMatrices(sigma_, DiscreteDistribution(sigma_), A_, B_, Pi_)\n",
    "    return [sigma_, hmmModel_]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# backup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "myGHMMDump(sigma, hmmModel, \"ghmmModel_standing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# restore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "[neverMind, hmmModel1] = myGHMMLoad(\"ghmmModel_standing\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 0, 0, 0, 0, 0], -31.77359208832549)\n"
     ]
    }
   ],
   "source": [
    "print(hmmModel1.viterbi(testSeq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on module ghmm:\n",
      "\n",
      "NAME\n",
      "    ghmm - @mainpage GHMM - an open source library for Hidden Markov Models (HMM)\n",
      "\n",
      "FILE\n",
      "    /home/kkuei/anaconda3/envs/rospy/lib/python2.7/site-packages/ghmm.py\n",
      "\n",
      "DESCRIPTION\n",
      "    HMMs are stochastic models which encode a probability density over\n",
      "    sequences of symbols. These symbols can be discrete letters (A,C,G and\n",
      "    T for DNA; 1,2,3,4,5,6 for dice), real numbers (weather measurement\n",
      "    over time: temperature) or vectors of either or the combination\n",
      "    thereof (weather again: temperature, pressure, percipitation).\n",
      "    \n",
      "    @note\n",
      "    We will always talk about emissions, emission sequence and so\n",
      "    forth when we refer to the sequence of symbols. Another name\n",
      "    for the same object is observation resp. observation sequence.\n",
      "    \n",
      "    A simple model with a fair and one unfair coin can be created as follows\n",
      "    \n",
      "    >> fair = [0.5, 0.5]\n",
      "    >> loaded = [0.9, 0.1]\n",
      "    >> A = [[0.9, 0.1], [0.3, 0.7]]\n",
      "    >> pi = [0.9, 0.1]\n",
      "    >> B = [fair, loaded]\n",
      "    >> sigma = ghmm.IntegerRange(0,2)\n",
      "    >> m = ghmm.HMMFromMatrices(sigma, ghmm.DiscreteDistribution(sigma), A, B, pi)\n",
      "    \n",
      "    The objects one has to deal with in HMM modelling are the following\n",
      "    \n",
      "    -# The domain the emissions come from: the EmissionDomain. Domain\n",
      "       is to be understood mathematically and to encompass both discrete,\n",
      "       finite alphabets and fields such as the real numbers or intervals\n",
      "       of the reals.\n",
      "    \n",
      "       For technical reasons there can be two representations of an\n",
      "       emission symbol: an external and an internal. The external\n",
      "       representation is the view of the application using ghmm.py. The\n",
      "       internal one is what is used in both ghmm.py and the ghmm\n",
      "       C-library. Representations can coincide, but this is not\n",
      "       guaranteed. Discrete alphabets of size k are represented as\n",
      "       [0,1,2,...,k-1] internally.  It is the domain objects job to\n",
      "       provide a mapping between representations in both directions.\n",
      "       @note\n",
      "       Do not make assumptions about the internal\n",
      "       representations. It might change.\n",
      "    \n",
      "    -# Every domain has to afford a distribution, which is usually\n",
      "       parameterized. A distribution associated with a domain\n",
      "       should allow us to compute \f",
      "$Prob[x| distribution parameters]\f",
      "$\n",
      "       efficiently.\n",
      "    \n",
      "       The distribution defines the type of distribution which\n",
      "       we will use to model emissions in <b>every state</b> of the HMM.\n",
      "       The type of distribution will be identical for all states,\n",
      "       their parameterizations will differ from state to state.\n",
      "    \n",
      "    -# We will consider a Sequence of emissions from the same emission\n",
      "       domain and very often sets of such sequences: SequenceSet\n",
      "    \n",
      "    -# The HMM: The HMM consists of two major components: A Markov chain\n",
      "       over states (implemented as a weighted directed graph with\n",
      "       adjacency and inverse-adjacency lists) and the emission\n",
      "       distributions per-state. For reasons of efficiency the HMM itself\n",
      "       is *static*, as far as the topology of the underlying Markov chain\n",
      "       (and obviously the EmissionDomain) are concerned. You cannot add or\n",
      "       delete transitions in an HMM.\n",
      "    \n",
      "       Transition probabilities and the parameters of the per-state\n",
      "       emission distributions can be easily modified. Particularly,\n",
      "       Baum-Welch reestimation is supported.  While a transition cannot be\n",
      "       deleted from the graph, you can set the transition probability to\n",
      "       zero, which has the same effect from the theoretical point of\n",
      "       view. However, the corresponding edge in the graph is still\n",
      "       traversed in the computation.\n",
      "    \n",
      "       States in HMMs are referred to by their integer index. State sequences\n",
      "       are simply list of integers.\n",
      "    \n",
      "       If you want to store application specific data for each state you\n",
      "       have to do it yourself.\n",
      "    \n",
      "       Subclasses of HMM implement specific types of HMM. The type depends\n",
      "       on the EmissionDomain, the Distribution used, the specific\n",
      "       extensions to the 'standard' HMMs and so forth\n",
      "\n",
      "CLASSES\n",
      "    __builtin__.object\n",
      "        BackgroundDistribution\n",
      "        ComplexEmissionSequence\n",
      "        Distribution\n",
      "            ContinuousDistribution\n",
      "                ContinuousMixtureDistribution\n",
      "                GaussianDistribution\n",
      "                    TruncGaussianDistribution\n",
      "                GaussianMixtureDistribution\n",
      "                MultivariateGaussianDistribution\n",
      "                UniformDistribution\n",
      "            DiscreteDistribution\n",
      "                DiscretePairDistribution\n",
      "        EmissionDomain\n",
      "            Alphabet\n",
      "                LabelDomain\n",
      "            Float\n",
      "        EmissionSequence\n",
      "        HMM\n",
      "            DiscreteEmissionHMM\n",
      "                StateLabelHMM\n",
      "            GaussianEmissionHMM\n",
      "                GaussianMixtureHMM\n",
      "                    ContinuousMixtureHMM\n",
      "                MultivariateGaussianMixtureHMM\n",
      "            PairHMM\n",
      "        HMMFactory\n",
      "            HMMFromMatricesFactory\n",
      "            HMMOpenFactory\n",
      "                PairHMMOpenFactory\n",
      "        SequenceSet\n",
      "            SequenceSetSubset\n",
      "    exceptions.Exception(exceptions.BaseException)\n",
      "        GHMMError\n",
      "            GHMMOutOfDomain\n",
      "            InvalidModelParameters\n",
      "            NoValidCDataType\n",
      "            ParseFileError\n",
      "            SequenceCannotBeBuild\n",
      "            UnknownInputType\n",
      "            UnsupportedFeature\n",
      "            WrongFileType\n",
      "            badCPointer\n",
      "    \n",
      "    class Alphabet(EmissionDomain)\n",
      "     |  Discrete, finite alphabet\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      Alphabet\n",
      "     |      EmissionDomain\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __eq__(self, alph)\n",
      "     |  \n",
      "     |  __hash__(self)\n",
      "     |  \n",
      "     |  __init__(self, listOfCharacters)\n",
      "     |      Creates an alphabet out of a listOfCharacters\n",
      "     |      @param listOfCharacters a list of strings (single characters most of\n",
      "     |      the time), ints, or other objects that can be used as dictionary keys\n",
      "     |      for a mapping of the external sequences to the internal representation\n",
      "     |      or can alternatively be a SWIG pointer to a\n",
      "     |      C alphabet_s struct\n",
      "     |      \n",
      "     |      @note\n",
      "     |      Alphabets should be considered as imutable. That means the\n",
      "     |      listOfCharacters and the mapping should never be touched after\n",
      "     |      construction.\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  external(self, internal)\n",
      "     |      Given an internal representation return the external representation\n",
      "     |      \n",
      "     |      @note the internal code -1 always represents a gap character '-'\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  externalSequence(self, internalSequence)\n",
      "     |      Given a sequence with the internal representation return the external\n",
      "     |      representation\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  getExternalCharacterLength(self)\n",
      "     |      If all external characters are of the same length the length is\n",
      "     |      returned. Otherwise None.\n",
      "     |      @return length of the external characters or None\n",
      "     |  \n",
      "     |  internal(self, emission)\n",
      "     |      Given a emission return the internal representation\n",
      "     |  \n",
      "     |  internalSequence(self, emissionSequence)\n",
      "     |      Given a emission_sequence return the internal representation\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  isAdmissable(self, emission)\n",
      "     |      Check whether emission is admissable (contained in) the domain\n",
      "     |  \n",
      "     |  size(self)\n",
      "     |      @deprecated use len() instead\n",
      "     |  \n",
      "     |  toCstruct(self)\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from EmissionDomain:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class BackgroundDistribution(__builtin__.object)\n",
      "     |  Background distributions object\n",
      "     |  \n",
      "     |  holds discrete distributions used as background while training\n",
      "     |  discrete HMMs to avoid overfitting.\n",
      "     |  Input is a discrete EmissionDomain and a list of list. Each list is\n",
      "     |  a distinct distribution. The distributions can be of higher order.\n",
      "     |  The length of a single distribution is a power of len(EmissionDomain)\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, bgInput)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  getCopy(self)\n",
      "     |  \n",
      "     |  getName(self, i)\n",
      "     |      return the name of the ith backgound distrubution\n",
      "     |  \n",
      "     |  setName(self, i, name)\n",
      "     |      sets the name of the ith background distrubution to name\n",
      "     |  \n",
      "     |  toLists(self)\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all background names to the dictionary name2id\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class ComplexEmissionSequence(__builtin__.object)\n",
      "     |  A MultivariateEmissionSequence is a sequence of multiple emissions per\n",
      "     |  time-point. Emissions can be from distinct EmissionDomains. In particular,\n",
      "     |  integer and floating point emissions are allowed. Access to emissions is\n",
      "     |  given by the index, seperately for discrete and continuous EmissionDomains.\n",
      "     |  \n",
      "     |  Example: XXX\n",
      "     |  \n",
      "     |  MultivariateEmissionSequence also links to the underlying C-structure.\n",
      "     |  \n",
      "     |  Note: ComplexEmissionSequence has to be considered imutable for the moment.\n",
      "     |  There are no means to manipulate the sequence positions yet.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation of C sequence struct.\n",
      "     |  \n",
      "     |  __getitem__(self, key)\n",
      "     |      get a slice of the complex emission sequence\n",
      "     |      @param key either int (makes no big sense) or slice object\n",
      "     |      @return a new ComplexEmissionSequence containing a slice of the\n",
      "     |      original\n",
      "     |  \n",
      "     |  __init__(self, emissionDomains, sequenceInputs, labelDomain=None, labelInput=None)\n",
      "     |      @param emissionDomains a list of EmissionDomain objects corresponding\n",
      "     |      to the list of sequenceInputs\n",
      "     |      @param sequenceInputs a list of sequences of the same length (e.g.\n",
      "     |      nucleotides and double values) that will be encoded\n",
      "     |      by the corresponding EmissionDomain\n",
      "     |      @bug @param labelDomain unused\n",
      "     |      @bug @param labelInput unused\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      @return the length of the sequence.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      string representation. Access the underlying C-structure and return\n",
      "     |      the sequence in all it's encodings (can be quite long)\n",
      "     |      @return string representation\n",
      "     |  \n",
      "     |  getDiscreteSequence(self, index)\n",
      "     |      get the 'index'th discrete sequence as it has been given at the input\n",
      "     |      @param index number of the discrete sequence\n",
      "     |      @return a python sequence\n",
      "     |  \n",
      "     |  getInternalContinuousSequence(self, index)\n",
      "     |      access the underlying C structure and return the internal\n",
      "     |      representation of the continuous sequence number 'index'\n",
      "     |      @param index number of the continuous sequence\n",
      "     |      @return a python list of floats\n",
      "     |  \n",
      "     |  getInternalDiscreteSequence(self, index)\n",
      "     |      access the underlying C structure and return the internal\n",
      "     |      representation of the discrete sequence number 'index'\n",
      "     |      @param index number of the discrete sequence\n",
      "     |      @return a python list of ints\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      string representation. Access the underlying C-structure and return\n",
      "     |      the sequence in all it's encodings (can be quite long)\n",
      "     |      @return string representation\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class ContinuousDistribution(Distribution)\n",
      "     |  Method resolution order:\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class ContinuousMixtureDistribution(ContinuousDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      ContinuousMixtureDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |  \n",
      "     |  add(self, w, fix, distribution)\n",
      "     |  \n",
      "     |  check(self)\n",
      "     |  \n",
      "     |  get(self, i)\n",
      "     |  \n",
      "     |  set(self, index, w, fix, distribution)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class ContinuousMixtureHMM(GaussianMixtureHMM)\n",
      "     |  HMMs with mixtures of any univariate (one dimensional) Continuous\n",
      "     |  Distributions as emissions.\n",
      "     |  \n",
      "     |  Optional features:\n",
      "     |  - fixing mixture components in training\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      ContinuousMixtureHMM\n",
      "     |      GaussianMixtureHMM\n",
      "     |      GaussianEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      defines string representation\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |      It also returns the density type\n",
      "     |  \n",
      "     |  getEmission(self, i, comp)\n",
      "     |      @returns the paramenters of component 'comp' in state 'i'\n",
      "     |      - (type, mu,  sigma^2, weight)        - for a gaussian component\n",
      "     |      - (type, mu,  sigma^2, min,   weight) - for a right tail gaussian\n",
      "     |      - (type, mu,  sigma^2, max,   weight) - for a left  tail gaussian\n",
      "     |      - (type, max, mix,     weight)        - for a uniform\n",
      "     |  \n",
      "     |  setEmission(self, i, comp, distType, values)\n",
      "     |      Set the emission distribution parameters for a mixture component\n",
      "     |      of a single state.\n",
      "     |      \n",
      "     |      @param i index of a state\n",
      "     |      @param comp index of a mixture component\n",
      "     |      @param distType type of the distribution\n",
      "     |      @param values tuple (mu, sigma, a , weight) and is interpreted depending\n",
      "     |      on distType\n",
      "     |      - mu     - mean for normal, normal_approx, normal_right, normal_left\n",
      "     |      - mu     - max for uniform\n",
      "     |      - sigma  - standard deviation for normal, normal_approx, normal_right,\n",
      "     |        normal_left\n",
      "     |      - sigma  - min for uniform\n",
      "     |      - a      - cut-off normal_right and normal_left\n",
      "     |      - weight - always component weight\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      Human readable model description\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from GaussianMixtureHMM:\n",
      "     |  \n",
      "     |  getMixtureFix(self, state)\n",
      "     |  \n",
      "     |  setMixtureFix(self, state, flags)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from GaussianEmissionHMM:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      Result: the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimate the model parameters given the training_sequences.\n",
      "     |      \n",
      "     |      Perform at most nr_steps until the improvement in likelihood\n",
      "     |      is below likelihood_cutoff\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |      \n",
      "     |      Result: Final loglikelihood\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |      Delete the necessary temporary variables for Baum-Welch-reestimation\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps, loglikelihoodCutoff=0.0001)\n",
      "     |      Setup necessary temporary variables for Baum-Welch-reestimation.\n",
      "     |      \n",
      "     |      Use with baumWelchStep for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |      Compute one iteration of Baum Welch estimation.\n",
      "     |      \n",
      "     |      Use with baumWelchSetup for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      Result: the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmissionProbability(self, value, i)\n",
      "     |      @returns probability of emitting value in state i\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      @returns the probability of the transition from state i to state j.\n",
      "     |      Raises IndexError if the transition is not allowed\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emissionSequences using the forward algorithm.\n",
      "     |      \n",
      "     |      @param emissionSequences SequenceSet\n",
      "     |      \n",
      "     |      Result: log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  viterbi(self, emissionSequences)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      \n",
      "     |      Result: [q_0, ..., q_T] the viterbi-path of emission_sequences is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class DiscreteDistribution(Distribution)\n",
      "     |  A DiscreteDistribution over an Alphabet: The discrete distribution\n",
      "     |  is parameterized by the vectors of probabilities.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      DiscreteDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, alphabet)\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, prob_vector)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class DiscreteEmissionHMM(HMM)\n",
      "     |  HMMs with discrete emissions.\n",
      "     |  \n",
      "     |  Optional features:\n",
      "     |  - silent states\n",
      "     |  - higher order states\n",
      "     |  - parameter tying in training\n",
      "     |  - background probabilities in training\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      DiscreteEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  applyBackgrounds(self, backgroundWeight)\n",
      "     |      Apply the background distribution to the emission probabilities of states\n",
      "     |      which have been assigned one (usually in the editor and coded in the XML).\n",
      "     |      \n",
      "     |      applyBackground computes a convex combination of the emission probability\n",
      "     |      and the background\n",
      "     |      \n",
      "     |      @param backgroundWeight (within [0,1]) controls the background's\n",
      "     |      contribution for each state.\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |  \n",
      "     |  backwardTermination(self, emissionSequence, pybeta, scalingVector)\n",
      "     |      Result: the backward log probability of emissionSequence\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimates the model with the sequence in 'trainingSequences'.\n",
      "     |      \n",
      "     |      @note that training for models including silent states is not yet\n",
      "     |      supported.\n",
      "     |      \n",
      "     |      @param trainingSequences EmissionSequence or SequenceSet object\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  cfbGibbs(self, trainingSequences, pA, pB, pPi, R=-1, burnIn=100, seed=0)\n",
      "     |      Reestimates the model and returns a sampled state sequence\n",
      "     |      \n",
      "     |      @note uses gsl, silent states not supported\n",
      "     |      \n",
      "     |      @param seed int for random seed, 0 default \n",
      "     |      @param trainingSequences EmissionSequence or SequenceSet\n",
      "     |      @param pA prior count for transitions\n",
      "     |      @param pB prior count for emissions\n",
      "     |      @param pPI prior count for initial state\n",
      "     |      @param R length of uniform compression >0, works best for .5log(sqrt(T)) where T is length of seq\n",
      "     |      @param burnin number of iterations\n",
      "     |      @return set of sampled paths for each training sequence\n",
      "     |      @warning work in progress\n",
      "     |  \n",
      "     |  extendDurations(self, durationlist)\n",
      "     |      extend states with durations larger than one.\n",
      "     |      \n",
      "     |      @note this done by explicit state copying in C\n",
      "     |  \n",
      "     |  fbGibbs(self, trainingSequences, pA, pB, pPi, burnIn=100, seed=0)\n",
      "     |      Reestimates the model and returns a sampled state sequence\n",
      "     |      \n",
      "     |      @note uses gsl, silent states not supported\n",
      "     |      \n",
      "     |      @param seed int for random seed, 0 default \n",
      "     |      @param trainingSequences EmissionSequence\n",
      "     |      @param pA prior count for transitions\n",
      "     |      @param pB prior count for emissions\n",
      "     |      @param pPI prior count for initial state\n",
      "     |      @param burnin number of iterations\n",
      "     |      @return set of sampled paths for each training sequence\n",
      "     |      @warning work in progress\n",
      "     |  \n",
      "     |  getBackgroundAssignments(self)\n",
      "     |      Get the background assignments of all states\n",
      "     |      \n",
      "     |      '-1' -> no background\n",
      "     |  \n",
      "     |  getEmission(self, i)\n",
      "     |  \n",
      "     |  getSilentFlag(self, state)\n",
      "     |  \n",
      "     |  getTieGroups(self)\n",
      "     |      Gets tied emission group structure.\n",
      "     |  \n",
      "     |  isSilent(self, state)\n",
      "     |      @returns True if 'state' is silent, False otherwise\n",
      "     |  \n",
      "     |  removeTieGroups(self)\n",
      "     |      Removes all tied emission information.\n",
      "     |  \n",
      "     |  setBackgroundAssignments(self, stateBackground)\n",
      "     |      Change all the assignments of background distributions to states.\n",
      "     |      \n",
      "     |      Input is a list of background ids or '-1' for no background, or list of background names\n",
      "     |  \n",
      "     |  setBackgrounds(self, backgroundObject, stateBackground)\n",
      "     |      Configure model to use the background distributions in 'backgroundObject'.\n",
      "     |      \n",
      "     |      @param backgroundObject BackgroundDistribution\n",
      "     |      @param 'stateBackground' a list of indixes (one for each state) refering\n",
      "     |      to distributions in 'backgroundObject'.\n",
      "     |      \n",
      "     |      @note values in backgroundObject are deep copied into the model\n",
      "     |  \n",
      "     |  setEmission(self, i, distributionParameters)\n",
      "     |      Set the emission distribution parameters for a discrete model.\n",
      "     |  \n",
      "     |  setTieGroups(self, tieList)\n",
      "     |      Sets the tied emission groups\n",
      "     |      \n",
      "     |      @param tieList contains for every state either '-1' or the index\n",
      "     |      of the tied emission group leader.\n",
      "     |      \n",
      "     |      @note The tied emission group leader is tied to itself\n",
      "     |  \n",
      "     |  updateTiedEmissions(self)\n",
      "     |      Averages emission probabilities of tied states.\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      @returns the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps)\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      @returns the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emission_sequences using the forward algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences is of type SequenceSet\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij.\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  viterbi(self, eseqs)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param eseqs can either be a SequenceSet or an EmissionSequence\n",
      "     |      \n",
      "     |      @returns [q_0, ..., q_T] the viterbi-path of \\p eseqs is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class DiscretePairDistribution(DiscreteDistribution)\n",
      "     |  A DiscreteDistribution over TWO Alphabets: The discrete distribution\n",
      "     |  is parameterized by the vector of probabilities.\n",
      "     |  To get the index of the vector that corresponds to a pair of characters\n",
      "     |  use the getPairIndex method.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      DiscretePairDistribution\n",
      "     |      DiscreteDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, alphabetX, alphabetY, offsetX, offsetY)\n",
      "     |      construct a new DiscretePairDistribution\n",
      "     |      @param alphabetX Alphabet object for sequence X\n",
      "     |      @param alphabetY Alphabet object for sequence Y\n",
      "     |      @param offsetX number of characters the alphabet of sequence X\n",
      "     |      consumes at a time\n",
      "     |      @param offsetY number of characters the alphabet of sequence Y\n",
      "     |      consumes at a time\n",
      "     |  \n",
      "     |  getCounts(self, sequenceX, sequenceY)\n",
      "     |      extract the pair counts for aligned sequences sequenceX and sequenceY\n",
      "     |      @param sequenceX string for sequence X\n",
      "     |      @param sequenceY strinf for sequence Y\n",
      "     |      @return a list of counts\n",
      "     |  \n",
      "     |  getEmptyProbabilityVector(self)\n",
      "     |      get an empty probability vector for this distribution (filled with 0.0)\n",
      "     |      @return list of floats\n",
      "     |  \n",
      "     |  getPairIndex(self, charX, charY)\n",
      "     |      get the index of a pair of two characters in the probability vector\n",
      "     |      (if you use the int representation both values must be ints)\n",
      "     |      @param charX character chain or int representation\n",
      "     |      @param charY character chain or int representation\n",
      "     |      @return the index of the pair in the probability vector\n",
      "     |  \n",
      "     |  setPairProbability(self, charX, charY, probability)\n",
      "     |      set the probability of the [air charX and charY to probability\n",
      "     |      @param charX character chain or int representation\n",
      "     |      @param charY character chain or int representation\n",
      "     |      @param probability probability (0<=float<=1)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from DiscreteDistribution:\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, prob_vector)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class Distribution(__builtin__.object)\n",
      "     |  Abstract base class for distribution over EmissionDomains\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class EmissionDomain(__builtin__.object)\n",
      "     |  Abstract base class for emissions produced by an HMM.\n",
      "     |  \n",
      "     |  There can be two representations for emissions:\n",
      "     |  -# An internal, used in ghmm.py and the ghmm C-library\n",
      "     |  -# An external, used in your particular application\n",
      "     |  \n",
      "     |  Example:\n",
      "     |  \n",
      "     |  The underlying library represents symbols from a finite,\n",
      "     |  discrete domain as integers (see Alphabet).\n",
      "     |  \n",
      "     |  EmissionDomain is the identity mapping\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  external(self, internal)\n",
      "     |      Given an internal representation return the external representation\n",
      "     |  \n",
      "     |  externalSequence(self, internalSequence)\n",
      "     |      Given a sequence with the internal representation return the external\n",
      "     |      representation\n",
      "     |  \n",
      "     |  internal(self, emission)\n",
      "     |      Given a emission return the internal representation\n",
      "     |  \n",
      "     |  internalSequence(self, emissionSequence)\n",
      "     |      Given a emissionSequence return the internal representation\n",
      "     |  \n",
      "     |  isAdmissable(self, emission)\n",
      "     |      Check whether \\p emission is admissable (contained in) the domain\n",
      "     |      raises GHMMOutOfDomain else\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class EmissionSequence(__builtin__.object)\n",
      "     |  An EmissionSequence contains the *internal* representation of\n",
      "     |  a sequence of emissions.\n",
      "     |  \n",
      "     |  It also contains a reference to the domain where the emissions orginated from.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation of C sequence struct.\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |      @returns the symbol at position 'index'.\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, sequenceInput, labelDomain=None, labelInput=None, ParentSequenceSet=None)\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Returns the length of the sequence.\n",
      "     |  \n",
      "     |  __setitem__(self, index, value)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  asSequenceSet(self)\n",
      "     |      @returns this EmissionSequence as a one element SequenceSet\n",
      "     |  \n",
      "     |  getGeneratingStates(self)\n",
      "     |      @returns the state path from which the sequence was generated as\n",
      "     |      a Python list.\n",
      "     |  \n",
      "     |  getSeqLabel(self)\n",
      "     |  \n",
      "     |  getStateLabel(self)\n",
      "     |      @returns the labeling of the sequence in external representation\n",
      "     |  \n",
      "     |  getWeight(self)\n",
      "     |  \n",
      "     |  hasStateLabels(self)\n",
      "     |      @returns whether the sequence is labeled or not\n",
      "     |  \n",
      "     |  sequenceSet(self)\n",
      "     |      @return a one-element SequenceSet with this sequence.\n",
      "     |  \n",
      "     |  setSeqLabel(self, value)\n",
      "     |  \n",
      "     |  setWeight(self, value)\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes the EmissionSequence into file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class Float(EmissionDomain)\n",
      "     |  Continuous Alphabet\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      Float\n",
      "     |      EmissionDomain\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __eq__(self, other)\n",
      "     |  \n",
      "     |  __hash__(self)\n",
      "     |  \n",
      "     |  __init__(self)\n",
      "     |  \n",
      "     |  isAdmissable(self, emission)\n",
      "     |      Check whether emission is admissable (contained in) the domain\n",
      "     |      \n",
      "     |      raises GHMMOutOfDomain else\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from EmissionDomain:\n",
      "     |  \n",
      "     |  external(self, internal)\n",
      "     |      Given an internal representation return the external representation\n",
      "     |  \n",
      "     |  externalSequence(self, internalSequence)\n",
      "     |      Given a sequence with the internal representation return the external\n",
      "     |      representation\n",
      "     |  \n",
      "     |  internal(self, emission)\n",
      "     |      Given a emission return the internal representation\n",
      "     |  \n",
      "     |  internalSequence(self, emissionSequence)\n",
      "     |      Given a emissionSequence return the internal representation\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from EmissionDomain:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class GHMMError(exceptions.Exception)\n",
      "     |  Base class for exceptions in this module.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class GHMMOutOfDomain(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      GHMMOutOfDomain\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class GaussianDistribution(ContinuousDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      GaussianDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |      # XXX attributes unused at this point\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, values)\n",
      "     |      @param values tuple of mu, sigma, trunc\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class GaussianEmissionHMM(HMM)\n",
      "     |  HMMs with Gaussian distribution as emissions.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      GaussianEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      Result: the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimate the model parameters given the training_sequences.\n",
      "     |      \n",
      "     |      Perform at most nr_steps until the improvement in likelihood\n",
      "     |      is below likelihood_cutoff\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |      \n",
      "     |      Result: Final loglikelihood\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |      Delete the necessary temporary variables for Baum-Welch-reestimation\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps, loglikelihoodCutoff=0.0001)\n",
      "     |      Setup necessary temporary variables for Baum-Welch-reestimation.\n",
      "     |      \n",
      "     |      Use with baumWelchStep for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |      Compute one iteration of Baum Welch estimation.\n",
      "     |      \n",
      "     |      Use with baumWelchSetup for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      Result: the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmission(self, i)\n",
      "     |      @returns (mu, sigma^2)\n",
      "     |  \n",
      "     |  getEmissionProbability(self, value, i)\n",
      "     |      @returns probability of emitting value in state i\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      @returns the probability of the transition from state i to state j.\n",
      "     |      Raises IndexError if the transition is not allowed\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emissionSequences using the forward algorithm.\n",
      "     |      \n",
      "     |      @param emissionSequences SequenceSet\n",
      "     |      \n",
      "     |      Result: log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  setEmission(self, i, values)\n",
      "     |      Set the emission distributionParameters for state i\n",
      "     |      \n",
      "     |      @param i index of a state\n",
      "     |      @param values tuple of mu, sigma\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  viterbi(self, emissionSequences)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      \n",
      "     |      Result: [q_0, ..., q_T] the viterbi-path of emission_sequences is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class GaussianMixtureDistribution(ContinuousDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      GaussianMixtureDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |      # XXX attributes unused at this point\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, index, values)\n",
      "     |      @param index index of mixture component\n",
      "     |      @param values tuple of mu, sigma, w\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class GaussianMixtureHMM(GaussianEmissionHMM)\n",
      "     |  HMMs with mixtures of Gaussians as emissions.\n",
      "     |  \n",
      "     |  Optional features:\n",
      "     |  - fixing mixture components in training\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      GaussianMixtureHMM\n",
      "     |      GaussianEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |  \n",
      "     |  getEmission(self, i, comp)\n",
      "     |      @returns (mu, sigma^2, weight) of component 'comp' in state 'i'\n",
      "     |  \n",
      "     |  getMixtureFix(self, state)\n",
      "     |  \n",
      "     |  setEmission(self, i, comp, values)\n",
      "     |      Set the emission distribution parameters for a single component in a single state.\n",
      "     |      \n",
      "     |      @param i index of a state\n",
      "     |      @param comp index of a mixture component\n",
      "     |      @param values tuple of mu, sigma, weight\n",
      "     |  \n",
      "     |  setMixtureFix(self, state, flags)\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      defines string representation\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from GaussianEmissionHMM:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      Result: the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimate the model parameters given the training_sequences.\n",
      "     |      \n",
      "     |      Perform at most nr_steps until the improvement in likelihood\n",
      "     |      is below likelihood_cutoff\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |      \n",
      "     |      Result: Final loglikelihood\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |      Delete the necessary temporary variables for Baum-Welch-reestimation\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps, loglikelihoodCutoff=0.0001)\n",
      "     |      Setup necessary temporary variables for Baum-Welch-reestimation.\n",
      "     |      \n",
      "     |      Use with baumWelchStep for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |      Compute one iteration of Baum Welch estimation.\n",
      "     |      \n",
      "     |      Use with baumWelchSetup for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      Result: the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmissionProbability(self, value, i)\n",
      "     |      @returns probability of emitting value in state i\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      @returns the probability of the transition from state i to state j.\n",
      "     |      Raises IndexError if the transition is not allowed\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emissionSequences using the forward algorithm.\n",
      "     |      \n",
      "     |      @param emissionSequences SequenceSet\n",
      "     |      \n",
      "     |      Result: log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  viterbi(self, emissionSequences)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      \n",
      "     |      Result: [q_0, ..., q_T] the viterbi-path of emission_sequences is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class HMM(__builtin__.object)\n",
      "     |  The HMM base class.\n",
      "     |  \n",
      "     |  All functions where the C signatures allows it will be defined in here.\n",
      "     |  Unfortunately there stil is a lot of overloading going on in derived classes.\n",
      "     |  \n",
      "     |  Generic features (these apply to all derived classes):\n",
      "     |  - Forward algorithm\n",
      "     |  - Viterbi algorithm\n",
      "     |  - Baum-Welch training\n",
      "     |  - HMM distance metric\n",
      "     |  - ...\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      To be defined in derived classes.\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      @returns the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      # The functions for model training are defined in the derived classes.\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps)\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      @returns the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmission(self, i)\n",
      "     |      Accessor function for the emission distribution parameters of state 'i'.\n",
      "     |      \n",
      "     |      For discrete models the distribution over the symbols is returned,\n",
      "     |      for continuous models a matrix of the form\n",
      "     |      [ [mu_1, sigma_1, weight_1] ... [mu_M, sigma_M, weight_M]  ] is returned.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emission_sequences using the forward algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences is of type SequenceSet\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setEmission(self, i, distributionParemters)\n",
      "     |      Set the emission distribution parameters\n",
      "     |      \n",
      "     |      Defined in derived classes.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij.\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  viterbi(self, eseqs)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param eseqs can either be a SequenceSet or an EmissionSequence\n",
      "     |      \n",
      "     |      @returns [q_0, ..., q_T] the viterbi-path of \\p eseqs is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class HMMFactory(__builtin__.object)\n",
      "     |  A HMMFactory is the base class of HMM factories.\n",
      "     |  A HMMFactory has just a constructor and a call method\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class HMMFromMatricesFactory(HMMFactory)\n",
      "     |  @todo Document matrix formats\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      HMMFromMatricesFactory\n",
      "     |      HMMFactory\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, emissionDomain, distribution, A, B, pi, hmmName=None, labelDomain=None, labelList=None, densities=None)\n",
      "     |      # XXX TODO: this should use the editing context\n",
      "     |  \n",
      "     |  constructSwitchingTransitions(self, cmodel, pi, A)\n",
      "     |      @internal function: creates switching transitions\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMMFactory:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class HMMOpenFactory(HMMFactory)\n",
      "     |  Opens a HMM from a file.\n",
      "     |  \n",
      "     |  Currently four formats are supported:\n",
      "     |  HMMer, our smo file format, and two xml formats.\n",
      "     |  @note the support for smo files and the old xml format will phase out\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      HMMOpenFactory\n",
      "     |      HMMFactory\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, fileName, modelIndex=None, filetype=None)\n",
      "     |  \n",
      "     |  __init__(self, defaultFileType=None)\n",
      "     |  \n",
      "     |  determineHMMClass(self, fileName)\n",
      "     |  \n",
      "     |  guessFileType(self, filename)\n",
      "     |      guesses the file format from the filename\n",
      "     |  \n",
      "     |  openHMMER(self, fileName)\n",
      "     |      Reads a file containing multiple HMMs in HMMER format, returns list of\n",
      "     |      HMM objects or a single HMM object.\n",
      "     |  \n",
      "     |  openNewXML(self, fileName, modelIndex)\n",
      "     |      Open one ore more HMM in the new xml format\n",
      "     |  \n",
      "     |  openOldXML(self, fileName)\n",
      "     |      #obsolete\n",
      "     |  \n",
      "     |  openSMO(self, fileName, modelIndex)\n",
      "     |      #obsolete\n",
      "     |  \n",
      "     |  openSingleHMMER(self, fileName)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMMFactory:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class InvalidModelParameters(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      InvalidModelParameters\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class LabelDomain(Alphabet)\n",
      "     |  # To be used for labelled HMMs. We could use an Alphabet directly but this way it is more explicit.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LabelDomain\n",
      "     |      Alphabet\n",
      "     |      EmissionDomain\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, listOfLabels)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from Alphabet:\n",
      "     |  \n",
      "     |  __eq__(self, alph)\n",
      "     |  \n",
      "     |  __hash__(self)\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  external(self, internal)\n",
      "     |      Given an internal representation return the external representation\n",
      "     |      \n",
      "     |      @note the internal code -1 always represents a gap character '-'\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  externalSequence(self, internalSequence)\n",
      "     |      Given a sequence with the internal representation return the external\n",
      "     |      representation\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  getExternalCharacterLength(self)\n",
      "     |      If all external characters are of the same length the length is\n",
      "     |      returned. Otherwise None.\n",
      "     |      @return length of the external characters or None\n",
      "     |  \n",
      "     |  internal(self, emission)\n",
      "     |      Given a emission return the internal representation\n",
      "     |  \n",
      "     |  internalSequence(self, emissionSequence)\n",
      "     |      Given a emission_sequence return the internal representation\n",
      "     |      \n",
      "     |      Raises KeyError\n",
      "     |  \n",
      "     |  isAdmissable(self, emission)\n",
      "     |      Check whether emission is admissable (contained in) the domain\n",
      "     |  \n",
      "     |  size(self)\n",
      "     |      @deprecated use len() instead\n",
      "     |  \n",
      "     |  toCstruct(self)\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from EmissionDomain:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class MultivariateGaussianDistribution(ContinuousDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      MultivariateGaussianDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class MultivariateGaussianMixtureHMM(GaussianEmissionHMM)\n",
      "     |  HMMs with Multivariate Gaussian distribution as emissions.\n",
      "     |  \n",
      "     |  States can have multiple mixture components.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      MultivariateGaussianMixtureHMM\n",
      "     |      GaussianEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, cmodel)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |  \n",
      "     |  getEmission(self, i, m)\n",
      "     |      @returns mean and covariance matrix of component m in state i\n",
      "     |  \n",
      "     |  setEmission(self, i, m, values)\n",
      "     |      Set the emission distributionParameters for mixture component m in\n",
      "     |      state i\n",
      "     |      \n",
      "     |      @param i index of a state\n",
      "     |      @param m index of a mixture component\n",
      "     |      @param values tuple of mu, sigma\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from GaussianEmissionHMM:\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      Result: the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimate the model parameters given the training_sequences.\n",
      "     |      \n",
      "     |      Perform at most nr_steps until the improvement in likelihood\n",
      "     |      is below likelihood_cutoff\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |      \n",
      "     |      Result: Final loglikelihood\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |      Delete the necessary temporary variables for Baum-Welch-reestimation\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps, loglikelihoodCutoff=0.0001)\n",
      "     |      Setup necessary temporary variables for Baum-Welch-reestimation.\n",
      "     |      \n",
      "     |      Use with baumWelchStep for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |      \n",
      "     |      @param trainingSequences can either be a SequenceSet or a Sequence\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |      Compute one iteration of Baum Welch estimation.\n",
      "     |      \n",
      "     |      Use with baumWelchSetup for more control over the training, computing\n",
      "     |      diagnostics or doing noise-insertion\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      Result: the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmissionProbability(self, value, i)\n",
      "     |      @returns probability of emitting value in state i\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      @returns the probability of the transition from state i to state j.\n",
      "     |      Raises IndexError if the transition is not allowed\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emissionSequences using the forward algorithm.\n",
      "     |      \n",
      "     |      @param emissionSequences SequenceSet\n",
      "     |      \n",
      "     |      Result: log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  viterbi(self, emissionSequences)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      \n",
      "     |      Result: [q_0, ..., q_T] the viterbi-path of emission_sequences is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class NoValidCDataType(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      NoValidCDataType\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class PairHMM(HMM)\n",
      "     |  Pair HMMs with discrete emissions over multiple alphabets.\n",
      "     |  Optional features: continuous values for transition classes\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PairHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomains, distribution, cmodel)\n",
      "     |      create a new PairHMM object (this should only be done using the\n",
      "     |      factory: e.g model = PairHMMOpenXML(modelfile) )\n",
      "     |      @param emissionDomains list of EmissionDomain objects\n",
      "     |      @param distribution (not used) inherited from HMM\n",
      "     |      @param cmodel a swig pointer on the underlying C structure\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      string representation (more for debuging) shows the contents of the C\n",
      "     |      structure ghmm_dpmodel\n",
      "     |      @return string representation\n",
      "     |  \n",
      "     |  addEmissionDomains(self, emissionDomains)\n",
      "     |      add additional EmissionDomains that are not specified in the XML file.\n",
      "     |      This is used to add information for the transition classes.\n",
      "     |      @param emissionDomains a list of EmissionDomain objects\n",
      "     |  \n",
      "     |  checkEmissions(self, eps=1e-13)\n",
      "     |      checks the sum of emission probabilities in all states\n",
      "     |      @param eps precision (if the sum is > 1 - eps it passes)\n",
      "     |      @return 1 if the emission of all states sum to one, 0 otherwise\n",
      "     |  \n",
      "     |  checkTransitions(self, eps=1e-13)\n",
      "     |      checks the sum of outgoing transition probabilities for all states\n",
      "     |      @param eps precision (if the sum is > 1 - eps it passes)\n",
      "     |      @return 1 if the transitions of all states sum to one, 0 otherwise\n",
      "     |  \n",
      "     |  logP(self, complexEmissionSequenceX, complexEmissionSequenceY, path)\n",
      "     |      compute the log probability of two sequences X and Y and a path\n",
      "     |      @param complexEmissionSequenceX sequence X encoded as\n",
      "     |      ComplexEmissionSequence\n",
      "     |      @param complexEmissionSequenceY sequence Y encoded as\n",
      "     |      ComplexEmissionSequence\n",
      "     |      @param path the state path\n",
      "     |      @return log probability\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      string representation (more for debuging) shows the contents of the C\n",
      "     |      structure ghmm_dpmodel\n",
      "     |      @return string representation\n",
      "     |  \n",
      "     |  viterbi(self, complexEmissionSequenceX, complexEmissionSequenceY)\n",
      "     |      run the naive implementation of the Viterbi algorithm and\n",
      "     |      return the viterbi path and the log probability of the path\n",
      "     |      @param complexEmissionSequenceX sequence X encoded as ComplexEmissionSequence\n",
      "     |      @param complexEmissionSequenceY sequence Y encoded as ComplexEmissionSequence\n",
      "     |      @return (path, log_p)\n",
      "     |  \n",
      "     |  viterbiPropagate(self, complexEmissionSequenceX, complexEmissionSequenceY, startX=None, startY=None, stopX=None, stopY=None, startState=None, startLogp=None, stopState=None, stopLogp=None)\n",
      "     |      run the linear space implementation of the Viterbi algorithm and\n",
      "     |      return the viterbi path and the log probability of the path\n",
      "     |      @param complexEmissionSequenceX sequence X encoded as ComplexEmissionSequence\n",
      "     |      @param complexEmissionSequenceY sequence Y encoded as ComplexEmissionSequence\n",
      "     |      Optional parameters to run the algorithm only on a segment:\n",
      "     |      @param startX start index in X\n",
      "     |      @param startY start index in Y\n",
      "     |      @param stopX stop index in X\n",
      "     |      @param stopY stop index in Y\n",
      "     |      @param startState start the path in this state\n",
      "     |      @param stopState path ends in this state\n",
      "     |      @param startLogp initialize the start state with this log probability\n",
      "     |      @param stopLogp if known this is the logp of the partial path\n",
      "     |      @return (path, log_p)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      To be defined in derived classes.\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      @returns the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      # The functions for model training are defined in the derived classes.\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps)\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      @returns the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getEmission(self, i)\n",
      "     |      Accessor function for the emission distribution parameters of state 'i'.\n",
      "     |      \n",
      "     |      For discrete models the distribution over the symbols is returned,\n",
      "     |      for continuous models a matrix of the form\n",
      "     |      [ [mu_1, sigma_1, weight_1] ... [mu_M, sigma_M, weight_M]  ] is returned.\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emission_sequences using the forward algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences is of type SequenceSet\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  sample(self, seqNr, T, seed=0)\n",
      "     |      Sample emission sequences.\n",
      "     |      \n",
      "     |      @param seqNr number of sequences to be sampled\n",
      "     |      @param T maximal length of each sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a SequenceSet object.\n",
      "     |  \n",
      "     |  sampleSingle(self, T, seed=0)\n",
      "     |      Sample a single emission sequence of length at most T.\n",
      "     |      \n",
      "     |      @param T maximal length of the sequence\n",
      "     |      @param seed initialization value for rng, default 0 leaves the state\n",
      "     |      of the rng alone\n",
      "     |      @returns a EmissionSequence object.\n",
      "     |  \n",
      "     |  setEmission(self, i, distributionParemters)\n",
      "     |      Set the emission distribution parameters\n",
      "     |      \n",
      "     |      Defined in derived classes.\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij.\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class PairHMMOpenFactory(HMMOpenFactory)\n",
      "     |  factory to create PairHMM objects from XML files\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PairHMMOpenFactory\n",
      "     |      HMMOpenFactory\n",
      "     |      HMMFactory\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, fileName_file_or_dom, modelIndex=None)\n",
      "     |      a call to the factory loads a model from a file specified by the\n",
      "     |      filename or from a file object or from a XML Document object and\n",
      "     |      initializes the model on the C side (libghmm).\n",
      "     |      @param fileName_file_or_dom load the model from a file specified by\n",
      "     |      a filename, a file object or a XML Document object\n",
      "     |      @param modelIndex not used (inherited from HMMOpenFactory)\n",
      "     |      @return PairHMM object\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMMOpenFactory:\n",
      "     |  \n",
      "     |  __init__(self, defaultFileType=None)\n",
      "     |  \n",
      "     |  determineHMMClass(self, fileName)\n",
      "     |  \n",
      "     |  guessFileType(self, filename)\n",
      "     |      guesses the file format from the filename\n",
      "     |  \n",
      "     |  openHMMER(self, fileName)\n",
      "     |      Reads a file containing multiple HMMs in HMMER format, returns list of\n",
      "     |      HMM objects or a single HMM object.\n",
      "     |  \n",
      "     |  openNewXML(self, fileName, modelIndex)\n",
      "     |      Open one ore more HMM in the new xml format\n",
      "     |  \n",
      "     |  openOldXML(self, fileName)\n",
      "     |      #obsolete\n",
      "     |  \n",
      "     |  openSMO(self, fileName, modelIndex)\n",
      "     |      #obsolete\n",
      "     |  \n",
      "     |  openSingleHMMER(self, fileName)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMMFactory:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class ParseFileError(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      ParseFileError\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class SequenceCannotBeBuild(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      SequenceCannotBeBuild\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class SequenceSet(__builtin__.object)\n",
      "     |  A SequenceSet contains the *internal* representation of a number of\n",
      "     |  sequences of emissions.\n",
      "     |  \n",
      "     |  It also contains a reference to the domain where the emissions orginated from.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation of C sequence struct.\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |      @returns an EmissionSequence object initialized with a reference to\n",
      "     |      sequence 'index'.\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, sequenceSetInput, labelDomain=None, labelInput=None)\n",
      "     |      @p sequenceSetInput is a set of sequences from @p emissionDomain.\n",
      "     |      \n",
      "     |      There are several valid types for @p sequenceSetInput:\n",
      "     |      - if @p sequenceSetInput is a string, it is interpreted as the filename\n",
      "     |        of a sequence file to be read. File format should be fasta.\n",
      "     |      - if @p sequenceSetInput is a list, it is considered as a list of lists\n",
      "     |        containing the input sequences\n",
      "     |      - @p sequenceSetInput can also be a pointer to a C sequence struct but\n",
      "     |        this is only meant for internal use\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      @returns the number of sequences in the SequenceSet.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  asSequenceSet(self)\n",
      "     |      convenience function, returns only self\n",
      "     |  \n",
      "     |  getGeneratingStates(self)\n",
      "     |      @returns the state paths from which the sequences were generated as a\n",
      "     |      Python list of lists.\n",
      "     |  \n",
      "     |  getSeqLabel(self, index)\n",
      "     |  \n",
      "     |  getSequence(self, index)\n",
      "     |      @returns the index-th sequence in internal representation\n",
      "     |  \n",
      "     |  getStateLabel(self, index)\n",
      "     |      @returns the labeling of the index-th sequence in internal representation\n",
      "     |  \n",
      "     |  getSubset(self, seqIndixes)\n",
      "     |      @returns a SequenceSet containing (references to) the sequences with the\n",
      "     |      indices in 'seqIndixes'.\n",
      "     |  \n",
      "     |  getWeight(self, i)\n",
      "     |      @returns the weight of sequence i. @note Weights are used in Baum-Welch\n",
      "     |  \n",
      "     |  hasStateLabels(self)\n",
      "     |      @returns whether the sequence is labeled or not\n",
      "     |  \n",
      "     |  merge(self, emissionSequences)\n",
      "     |      Merges 'emissionSequences' into 'self'.\n",
      "     |      @param emissionSequences can either be an EmissionSequence or SequenceSet\n",
      "     |      object.\n",
      "     |  \n",
      "     |  sequenceLength(self, i)\n",
      "     |      @returns the lenght of sequence 'i' in the SequenceSet\n",
      "     |  \n",
      "     |  setSeqLabel(self, index, value)\n",
      "     |  \n",
      "     |  setWeight(self, i, w)\n",
      "     |      Set the weight of sequence i. @note Weights are used in Baum-Welch\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes (appends) the SequenceSet into file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class SequenceSetSubset(SequenceSet)\n",
      "     |  SequenceSetSubset contains a subset of the sequences from a SequenceSet\n",
      "     |  object.\n",
      "     |  \n",
      "     |  @note On the C side only the references are used.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      SequenceSetSubset\n",
      "     |      SequenceSet\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Since we do not want to deallocate the sequence memory,\n",
      "     |      the destructor has to be overloaded.\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, sequenceSetInput, ParentSequenceSet, labelDomain=None, labelInput=None)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from SequenceSet:\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |      @returns an EmissionSequence object initialized with a reference to\n",
      "     |      sequence 'index'.\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      @returns the number of sequences in the SequenceSet.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  asSequenceSet(self)\n",
      "     |      convenience function, returns only self\n",
      "     |  \n",
      "     |  getGeneratingStates(self)\n",
      "     |      @returns the state paths from which the sequences were generated as a\n",
      "     |      Python list of lists.\n",
      "     |  \n",
      "     |  getSeqLabel(self, index)\n",
      "     |  \n",
      "     |  getSequence(self, index)\n",
      "     |      @returns the index-th sequence in internal representation\n",
      "     |  \n",
      "     |  getStateLabel(self, index)\n",
      "     |      @returns the labeling of the index-th sequence in internal representation\n",
      "     |  \n",
      "     |  getSubset(self, seqIndixes)\n",
      "     |      @returns a SequenceSet containing (references to) the sequences with the\n",
      "     |      indices in 'seqIndixes'.\n",
      "     |  \n",
      "     |  getWeight(self, i)\n",
      "     |      @returns the weight of sequence i. @note Weights are used in Baum-Welch\n",
      "     |  \n",
      "     |  hasStateLabels(self)\n",
      "     |      @returns whether the sequence is labeled or not\n",
      "     |  \n",
      "     |  merge(self, emissionSequences)\n",
      "     |      Merges 'emissionSequences' into 'self'.\n",
      "     |      @param emissionSequences can either be an EmissionSequence or SequenceSet\n",
      "     |      object.\n",
      "     |  \n",
      "     |  sequenceLength(self, i)\n",
      "     |      @returns the lenght of sequence 'i' in the SequenceSet\n",
      "     |  \n",
      "     |  setSeqLabel(self, index, value)\n",
      "     |  \n",
      "     |  setWeight(self, i, w)\n",
      "     |      Set the weight of sequence i. @note Weights are used in Baum-Welch\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |      Defines string representation.\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes (appends) the SequenceSet into file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from SequenceSet:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class StateLabelHMM(DiscreteEmissionHMM)\n",
      "     |  Labelled HMMs with discrete emissions.\n",
      "     |  \n",
      "     |  Same feature list as in DiscreteEmissionHMM models.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      StateLabelHMM\n",
      "     |      DiscreteEmissionHMM\n",
      "     |      HMM\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, emissionDomain, distribution, labelDomain, cmodel)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  externalLabel(self, internal)\n",
      "     |      @returns label representation of an int or list of ints\n",
      "     |  \n",
      "     |  getLabel(self, stateIndex)\n",
      "     |      @returns label of the state 'stateIndex'.\n",
      "     |  \n",
      "     |  getLabels(self)\n",
      "     |  \n",
      "     |  gradientSearch(self, emissionSequences, eta=0.1, steps=20)\n",
      "     |      trains a model with given sequences using a gradient descent algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      @param eta algortihm terminates if the descent is smaller than eta\n",
      "     |      @param steps number of iterations\n",
      "     |  \n",
      "     |  internalLabel(self, external)\n",
      "     |      @returns int representation of an label or list of labels\n",
      "     |  \n",
      "     |  kbest(self, emissionSequences, k=1)\n",
      "     |      Compute the k probable labeling for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or an\n",
      "     |      EmissionSequence\n",
      "     |      @param k the number of labelings to produce\n",
      "     |      \n",
      "     |      Result: [l_0, ..., l_T] the labeling of emissionSequences is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[l_0^0, ..., l_T^0], ..., [l_0^j, ..., l_T^j]} for a j-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  labeledBackward(self, emissionSequence, labelSequence, scalingVector)\n",
      "     |      Result: the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  labeledBaumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimates the model with the sequence in 'trainingSequences'.\n",
      "     |      \n",
      "     |      @note that training for models including silent states is not yet\n",
      "     |      supported.\n",
      "     |      \n",
      "     |      @param trainingSequences EmissionSequence or SequenceSet object\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  labeledForward(self, emissionSequence, labelSequence)\n",
      "     |      Result: the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  labeledViterbi(self, emissionSequences)\n",
      "     |      @returns the labeling of the input sequence(s) as given by the viterbi\n",
      "     |      path.\n",
      "     |      \n",
      "     |      For one EmissionSequence a list of labels is returned; for an SequenceSet\n",
      "     |      a list of lists of labels.\n",
      "     |  \n",
      "     |  labeledlogikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s,l| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual \\p emissionSequences using the forward algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences SequenceSet\n",
      "     |      \n",
      "     |      Result: log( P[emissionSequences,labels| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  sample(self, seqNr, seqLength, seed=0)\n",
      "     |  \n",
      "     |  sampleSingle(self, seqLength, seed=0)\n",
      "     |  \n",
      "     |  setLabels(self, labelList)\n",
      "     |      Set the state labels to the values given in labelList.\n",
      "     |      \n",
      "     |      LabelList is in external representation.\n",
      "     |  \n",
      "     |  verboseStr(self)\n",
      "     |  \n",
      "     |  write(self, fileName)\n",
      "     |      Writes HMM to file 'fileName'.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from DiscreteEmissionHMM:\n",
      "     |  \n",
      "     |  applyBackgrounds(self, backgroundWeight)\n",
      "     |      Apply the background distribution to the emission probabilities of states\n",
      "     |      which have been assigned one (usually in the editor and coded in the XML).\n",
      "     |      \n",
      "     |      applyBackground computes a convex combination of the emission probability\n",
      "     |      and the background\n",
      "     |      \n",
      "     |      @param backgroundWeight (within [0,1]) controls the background's\n",
      "     |      contribution for each state.\n",
      "     |  \n",
      "     |  asMatrices(self)\n",
      "     |      Return the parameters in matrix form.\n",
      "     |  \n",
      "     |  backwardTermination(self, emissionSequence, pybeta, scalingVector)\n",
      "     |      Result: the backward log probability of emissionSequence\n",
      "     |  \n",
      "     |  baumWelch(self, trainingSequences, nrSteps=500, loglikelihoodCutoff=0.0001)\n",
      "     |      Reestimates the model with the sequence in 'trainingSequences'.\n",
      "     |      \n",
      "     |      @note that training for models including silent states is not yet\n",
      "     |      supported.\n",
      "     |      \n",
      "     |      @param trainingSequences EmissionSequence or SequenceSet object\n",
      "     |      @param nrSteps the maximal number of BW-steps\n",
      "     |      @param loglikelihoodCutoff the least relative improvement in likelihood\n",
      "     |      with respect to the last iteration required to continue.\n",
      "     |  \n",
      "     |  cfbGibbs(self, trainingSequences, pA, pB, pPi, R=-1, burnIn=100, seed=0)\n",
      "     |      Reestimates the model and returns a sampled state sequence\n",
      "     |      \n",
      "     |      @note uses gsl, silent states not supported\n",
      "     |      \n",
      "     |      @param seed int for random seed, 0 default \n",
      "     |      @param trainingSequences EmissionSequence or SequenceSet\n",
      "     |      @param pA prior count for transitions\n",
      "     |      @param pB prior count for emissions\n",
      "     |      @param pPI prior count for initial state\n",
      "     |      @param R length of uniform compression >0, works best for .5log(sqrt(T)) where T is length of seq\n",
      "     |      @param burnin number of iterations\n",
      "     |      @return set of sampled paths for each training sequence\n",
      "     |      @warning work in progress\n",
      "     |  \n",
      "     |  extendDurations(self, durationlist)\n",
      "     |      extend states with durations larger than one.\n",
      "     |      \n",
      "     |      @note this done by explicit state copying in C\n",
      "     |  \n",
      "     |  fbGibbs(self, trainingSequences, pA, pB, pPi, burnIn=100, seed=0)\n",
      "     |      Reestimates the model and returns a sampled state sequence\n",
      "     |      \n",
      "     |      @note uses gsl, silent states not supported\n",
      "     |      \n",
      "     |      @param seed int for random seed, 0 default \n",
      "     |      @param trainingSequences EmissionSequence\n",
      "     |      @param pA prior count for transitions\n",
      "     |      @param pB prior count for emissions\n",
      "     |      @param pPI prior count for initial state\n",
      "     |      @param burnin number of iterations\n",
      "     |      @return set of sampled paths for each training sequence\n",
      "     |      @warning work in progress\n",
      "     |  \n",
      "     |  getBackgroundAssignments(self)\n",
      "     |      Get the background assignments of all states\n",
      "     |      \n",
      "     |      '-1' -> no background\n",
      "     |  \n",
      "     |  getEmission(self, i)\n",
      "     |  \n",
      "     |  getSilentFlag(self, state)\n",
      "     |  \n",
      "     |  getTieGroups(self)\n",
      "     |      Gets tied emission group structure.\n",
      "     |  \n",
      "     |  isSilent(self, state)\n",
      "     |      @returns True if 'state' is silent, False otherwise\n",
      "     |  \n",
      "     |  removeTieGroups(self)\n",
      "     |      Removes all tied emission information.\n",
      "     |  \n",
      "     |  setBackgroundAssignments(self, stateBackground)\n",
      "     |      Change all the assignments of background distributions to states.\n",
      "     |      \n",
      "     |      Input is a list of background ids or '-1' for no background, or list of background names\n",
      "     |  \n",
      "     |  setBackgrounds(self, backgroundObject, stateBackground)\n",
      "     |      Configure model to use the background distributions in 'backgroundObject'.\n",
      "     |      \n",
      "     |      @param backgroundObject BackgroundDistribution\n",
      "     |      @param 'stateBackground' a list of indixes (one for each state) refering\n",
      "     |      to distributions in 'backgroundObject'.\n",
      "     |      \n",
      "     |      @note values in backgroundObject are deep copied into the model\n",
      "     |  \n",
      "     |  setEmission(self, i, distributionParameters)\n",
      "     |      Set the emission distribution parameters for a discrete model.\n",
      "     |  \n",
      "     |  setTieGroups(self, tieList)\n",
      "     |      Sets the tied emission groups\n",
      "     |      \n",
      "     |      @param tieList contains for every state either '-1' or the index\n",
      "     |      of the tied emission group leader.\n",
      "     |      \n",
      "     |      @note The tied emission group leader is tied to itself\n",
      "     |  \n",
      "     |  updateTiedEmissions(self)\n",
      "     |      Averages emission probabilities of tied states.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from HMM:\n",
      "     |  \n",
      "     |  __del__(self)\n",
      "     |      Deallocation routine for the underlying C data structures.\n",
      "     |  \n",
      "     |  backward(self, emissionSequence, scalingVector)\n",
      "     |      @returns the (N x T)-matrix containing the backward-variables\n",
      "     |  \n",
      "     |  baumWelchDelete(self)\n",
      "     |  \n",
      "     |  baumWelchSetup(self, trainingSequences, nrSteps)\n",
      "     |  \n",
      "     |  baumWelchStep(self, nrSteps, loglikelihoodCutoff)\n",
      "     |  \n",
      "     |  clearFlags(self, flags)\n",
      "     |      Clears one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  distance(self, model, seqLength)\n",
      "     |      @returns the distance between 'self.cmodel' and 'model'.\n",
      "     |  \n",
      "     |  forward(self, emissionSequence)\n",
      "     |      @returns the (N x T)-matrix containing the forward-variables\n",
      "     |      and the scaling vector\n",
      "     |  \n",
      "     |  getInitial(self, i)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$\n",
      "     |  \n",
      "     |  getStateFix(self, state)\n",
      "     |  \n",
      "     |  getStateName(self, index)\n",
      "     |      returns the name of the state index\n",
      "     |  \n",
      "     |  getTransition(self, i, j)\n",
      "     |      Accessor function for the transition a_ij\n",
      "     |  \n",
      "     |  hasFlags(self, flags)\n",
      "     |      Checks if the model has one or more model type flags set\n",
      "     |  \n",
      "     |  joined(self, emissionSequence, stateSequence)\n",
      "     |      log P[ emissionSequence, stateSequence| m]\n",
      "     |  \n",
      "     |  loglikelihood(self, emissionSequences)\n",
      "     |      Compute log( P[emissionSequences| model]) using the forward algorithm\n",
      "     |      assuming independence of the sequences in emissionSequences\n",
      "     |      \n",
      "     |      @param emissionSequences can either be a SequenceSet or a EmissionSequence\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float which is\n",
      "     |      computed as \f",
      "$\\sum_{s} log( P[s| model])\f",
      "$ when emissionSequences\n",
      "     |      is a SequenceSet\n",
      "     |      \n",
      "     |      @note The implementation does not compute the full forward matrix since\n",
      "     |      we are only interested in the likelihoods in this case.\n",
      "     |  \n",
      "     |  loglikelihoods(self, emissionSequences)\n",
      "     |      Compute a vector ( log( P[s| model]) )_{s} of log-likelihoods of the\n",
      "     |      individual emission_sequences using the forward algorithm\n",
      "     |      \n",
      "     |      @param emissionSequences is of type SequenceSet\n",
      "     |      \n",
      "     |      @returns log( P[emissionSequences| model]) of type float\n",
      "     |      (numarray) vector of floats\n",
      "     |  \n",
      "     |  normalize(self)\n",
      "     |      Normalize transition probs, emission probs (if applicable)\n",
      "     |  \n",
      "     |  pathPosterior(self, sequence, path)\n",
      "     |      @returns the log posterior probability for 'path' having generated\n",
      "     |      'sequence'.\n",
      "     |      \n",
      "     |      @attention pathPosterior needs to calculate the complete forward and\n",
      "     |      backward matrices. If you are interested in multiple paths it would\n",
      "     |      be more efficient to use the 'posterior' function directly and not\n",
      "     |      multiple calls to pathPosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  posterior(self, sequence)\n",
      "     |      Posterior distribution matrix for 'sequence'.\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  printtypes(self, model_type)\n",
      "     |  \n",
      "     |  randomize(self, noiseLevel)\n",
      "     |      to be defined in derived class\n",
      "     |  \n",
      "     |  setFlags(self, flags)\n",
      "     |      Sets one or more model type flags.\n",
      "     |      @attention Use with care.\n",
      "     |  \n",
      "     |  setInitial(self, i, prob, fixProb=False)\n",
      "     |      Accessor function for the initial probability \f",
      "$\\pi_i\f",
      "$.\n",
      "     |      \n",
      "     |      If 'fixProb' = True \f",
      "$\\pi\f",
      "$ will be rescaled to 1 with 'pi[i]'\n",
      "     |      fixed to the arguement value of 'prob'.\n",
      "     |  \n",
      "     |  setStateFix(self, state, flag)\n",
      "     |  \n",
      "     |  setStateName(self, index, name)\n",
      "     |      sets the state name of state index to name\n",
      "     |  \n",
      "     |  setTransition(self, i, j, prob)\n",
      "     |      Accessor function for the transition a_ij.\n",
      "     |  \n",
      "     |  state(self, stateLabel)\n",
      "     |      Given a stateLabel return the integer index to the state\n",
      "     |  \n",
      "     |  statePosterior(self, sequence, state, time)\n",
      "     |      @returns the log posterior probability for being at 'state'\n",
      "     |      at time 'time' in 'sequence'.\n",
      "     |      \n",
      "     |      @attention: statePosterior needs to calculate the complete forward\n",
      "     |      and backward matrices. If you are interested in multiple states\n",
      "     |      it would be more efficient to use the posterior function directly\n",
      "     |      and not multiple calls to statePosterior\n",
      "     |      \n",
      "     |      @todo for silent states things are more complicated -> to be done\n",
      "     |  \n",
      "     |  updateName2id(self)\n",
      "     |      adds all state names to the dictionary name2id\n",
      "     |  \n",
      "     |  viterbi(self, eseqs)\n",
      "     |      Compute the Viterbi-path for each sequence in emissionSequences\n",
      "     |      \n",
      "     |      @param eseqs can either be a SequenceSet or an EmissionSequence\n",
      "     |      \n",
      "     |      @returns [q_0, ..., q_T] the viterbi-path of \\p eseqs is an\n",
      "     |      EmmissionSequence object,\n",
      "     |      [[q_0^0, ..., q_T^0], ..., [q_0^k, ..., q_T^k]} for a k-sequence\n",
      "     |      SequenceSet\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HMM:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class TruncGaussianDistribution(GaussianDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      TruncGaussianDistribution\n",
      "     |      GaussianDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |      # XXX attributes unused at this point\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, values)\n",
      "     |      @param values tuple of mu, sigma, trunc\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class UniformDistribution(ContinuousDistribution)\n",
      "     |  Method resolution order:\n",
      "     |      UniformDistribution\n",
      "     |      ContinuousDistribution\n",
      "     |      Distribution\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, domain)\n",
      "     |  \n",
      "     |  get(self)\n",
      "     |  \n",
      "     |  set(self, values)\n",
      "     |      @param values tuple of maximum, minimum\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from Distribution:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class UnknownInputType(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      UnknownInputType\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class UnsupportedFeature(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      UnsupportedFeature\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class WrongFileType(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      WrongFileType\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "    \n",
      "    class badCPointer(GHMMError)\n",
      "     |  Method resolution order:\n",
      "     |      badCPointer\n",
      "     |      GHMMError\n",
      "     |      exceptions.Exception\n",
      "     |      exceptions.BaseException\n",
      "     |      __builtin__.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, message)\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from GHMMError:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from exceptions.Exception:\n",
      "     |  \n",
      "     |  __new__ = <built-in method __new__ of type object>\n",
      "     |      T.__new__(S, ...) -> a new object with type S, a subtype of T\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(...)\n",
      "     |      x.__delattr__('name') <==> del x.name\n",
      "     |  \n",
      "     |  __getattribute__(...)\n",
      "     |      x.__getattribute__('name') <==> x.name\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __getslice__(...)\n",
      "     |      x.__getslice__(i, j) <==> x[i:j]\n",
      "     |      \n",
      "     |      Use of negative indices is not supported.\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |  \n",
      "     |  __repr__(...)\n",
      "     |      x.__repr__() <==> repr(x)\n",
      "     |  \n",
      "     |  __setattr__(...)\n",
      "     |      x.__setattr__('name', value) <==> x.name = value\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __unicode__(...)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from exceptions.BaseException:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  args\n",
      "     |  \n",
      "     |  message\n",
      "\n",
      "FUNCTIONS\n",
      "    HMMDiscriminativePerformance(HMMList, SeqList)\n",
      "        Computes the discriminative performce of the HMMs in HMMList\n",
      "        under the sequences in SeqList\n",
      "    \n",
      "    HMMDiscriminativeTraining(HMMList, SeqList, nrSteps=50, gradient=0)\n",
      "        Trains a couple of HMMs to increase the probablistic distance\n",
      "        if the the HMMs are used as classifier.\n",
      "        \n",
      "        @param HMMList List of labeled HMMs\n",
      "        @param SeqList List of labeled sequences, one for each HMM\n",
      "        @param nrSteps maximal number of iterations\n",
      "        @param gradient @todo document me\n",
      "        \n",
      "        @note this method does a initial expectation maximization training\n",
      "    \n",
      "    HMMwriteList(fileName, hmmList, fileType='xml')\n",
      "    \n",
      "    IntegerRange(a, b)\n",
      "        Creates an Alphabet with internal and external representation of range(a,b)\n",
      "        @return Alphabet\n",
      "    \n",
      "    SequenceSetOpen(emissionDomain, fileName)\n",
      "        Reads a sequence file with multiple sequence sets.\n",
      "        \n",
      "        @returns a list of SequenceSet objects.\n",
      "    \n",
      "    logwrapper(level, message)\n",
      "    \n",
      "    writeToFasta(seqSet, fn)\n",
      "        Writes a SequenceSet into a fasta file.\n",
      "\n",
      "DATA\n",
      "    AminoAcids = <ghmm.Alphabet object>\n",
      "    DNA = <ghmm.Alphabet object>\n",
      "    GHMM_FILETYPE_HMMER = 'hmm'\n",
      "    GHMM_FILETYPE_SMO = 'smo'\n",
      "    GHMM_FILETYPE_XML = 'xml'\n",
      "    HMMFromMatrices = <ghmm.HMMFromMatricesFactory object>\n",
      "    HMMOpen = <ghmm.HMMOpenFactory object>\n",
      "    HMMOpenHMMER = <ghmm.HMMOpenFactory object>\n",
      "    HMMOpenSMO = <ghmm.HMMOpenFactory object>\n",
      "    HMMOpenXML = <ghmm.HMMOpenFactory object>\n",
      "    PairHMMOpenXML = <ghmm.PairHMMOpenFactory object>\n",
      "    c_log = [<bound method Logger.critical of <logging.Logger object>>, <b...\n",
      "    fmt = <logging.Formatter object>\n",
      "    hdlr = <logging.StreamHandler object>\n",
      "    kBackgroundDistributions = 32\n",
      "    kContinuousHMM = 512\n",
      "    kDiscreteHMM = 256\n",
      "    kHigherOrderEmissions = 16\n",
      "    kLabeledStates = 64\n",
      "    kLeftRight = 1\n",
      "    kNotSpecified = 0\n",
      "    kPairHMM = 1024\n",
      "    kSilentStates = 4\n",
      "    kTiedEmissions = 8\n",
      "    kTransitionClasses = 128\n",
      "    log = <logging.Logger object>\n",
      "    types = {1: 'kLeftRight', 4: 'kSilentStates', 8: 'kTiedEmissions', 16:...\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help('ghmm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:rospy]",
   "language": "python",
   "name": "conda-env-rospy-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
